### PR TITLE
Improve Two Way Sync Behaviour

### PR DESCRIPTION
**Pull request by alperozturk96**
- [x] Tests written, or not not needed

• Updates getInternalTwoWaySyncFolders() to return only folders, excluding the root directory, avoiding redundant checks in the worker.
• Adds a toggle to enable/disable two-way sync.
• Adds a dropdown menu to adjust the two-way sync interval.

Screenshot 2024-10-28 at 12 45 23

### CODE REVIEW

File: app/src/main/java/com/nextcloud/client/database/dao/FileDao.kt [hunk 1]
```diff
@@ -55,4 +55,10 @@ interface FileDao {
             "ORDER BY internal_two_way_sync_timestamp DESC"
     )
     fun getInternalTwoWaySyncFolders(fileOwner: String): List<FileEntity>
+
+    @Query(
+        "UPDATE filelist SET internal_two_way_sync_timestamp = -1 WHERE internal_two_way_sync_timestamp " +
+            "IS NULL OR internal_two_way_sync_timestamp >= 0"
+    )
+    fun setInternalTwoWaySyncTimestampToMinusOne()
```
Comment (solonovamax) [inline app/src/main/java/com/nextcloud/client/database/dao/FileDao.kt:63]: could this not be better named, such as `resetTwoWaySyncTimestamp()`?

File: app/src/main/java/com/nextcloud/client/jobs/InternalTwoWaySyncWork.kt [hunk 2]
```diff
@@ -54,13 +58,6 @@ class InternalTwoWaySyncWork(
                     return checkFreeSpaceResult
                 }
 
-                // do not attempt to sync root folder
-                if (folder.remotePath == OCFile.ROOT_PATH) {
-                    folder.internalFolderSyncTimestamp = -1L
-                    fileDataStorageManager.saveFile(folder)
-                    continue
-                }
-
```
Comment (solonovamax) [inline app/src/main/java/com/nextcloud/client/jobs/InternalTwoWaySyncWork.kt:63]: not familiar with the nextcloud codebase, but why was this removed? this seems important, and I don't see how removing it might affect things.

but, I could be wrong. as I said, not familiar with the codebase.

File: app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java [hunk 3]
```diff
@@ -789,4 +793,34 @@ public void setLastSelectedMediaFolder(@NonNull String path) {
...
+
+    @Override
+    public boolean isTwoWaySyncEnabled() {
+        return preferences.getBoolean(PREF__TWO_WAY_STATUS, true);
+    }
+
+    @Override
+    public void setTwoWaySyncTimestampDefault(boolean value) {
+        preferences.edit().putBoolean(PREF__RESET_TWO_WAY_SYNC_TIMESTAMP, value).apply();
+    }
```
Comment (solonovamax) [inline app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java:810]: why is it called "two way sync timestamp default" in some places (eg. this method), and "reset two way sync timestamp" (eg. the preference) in others?

File: app/src/main/java/com/nextcloud/model/DurationOption.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+package com.nextcloud.model
+
+import android.content.Context
+import com.owncloud.android.R
+
+data class DurationOption(
+    val value: Long,
```
Comment (solonovamax) [inline app/src/main/java/com/nextcloud/model/DurationOption.kt:14]: might be a good idea to use a kotlin [Duration](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/) here?

File: app/src/main/java/com/nextcloud/client/jobs/InternalTwoWaySyncWork.kt (hunk 2 above)
Comment (alperozturk96) [reply app/src/main/java/com/nextcloud/client/jobs/InternalTwoWaySyncWork.kt:63]: Instead of checking the root folder inside the worker, I’m passing the correct folders directly to it, so there’s no need for checks within the worker.


See changes: https://github.com/nextcloud/android/pull/13884/commits/f1726bece60420f450ea081ca3b68b0b8be850af

File: app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java (hunk 3 above)
Comment (alperozturk96) [reply app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java:810]: This method is used to ensure that we update the incorrect column value only once. PR still under development, feel free to suggest function names, thank you.

File: app/src/main/java/com/nextcloud/model/DurationOption.kt (hunk 4 above)
Comment (alperozturk96) [reply app/src/main/java/com/nextcloud/model/DurationOption.kt:14]: Ahh cool thanks 💯 

.toIsoString() // returns "PT15M" 

Can we get 15 minutes, 30 minutes, 1 hour, 2 hours ... string representation from kotlin.Duration? I want to get rid of also translations if kotlin.Duration can provide it.

File: app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java (hunk 3 above)
Comment (alperozturk96) [reply app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java]: [remove reset column function due to "remove all" button will be added in three dot menu](https://github.com/nextcloud/android/pull/13884/commits/76ca0e6b736469c3dbe81790edc47b0c7321eaf3)

**Review by ZetaTom — COMMENTED** [review]
Everything seems to work during testing. Suggestions are mostly UI improvements.

(However, the worker still re-downloads everything on every run. This should be fixed by #13897.)

File: app/src/main/res/layout/internal_two_way_sync_layout.xml [hunk 5]
```diff
@@ -5,14 +5,60 @@
...
+            android:layout_height="wrap_content"/>
+
+        <com.google.android.material.materialswitch.MaterialSwitch
+            android:id="@+id/twoWaySyncToggle"
+            android:scaleX="0.7"
+            android:scaleY="0.7"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content" />
+
+    </LinearLayout>
```
[the remark this answers is not in the document]
Comment (alperozturk96) [reply app/src/main/res/layout/internal_two_way_sync_layout.xml:38]: Exactly, it was MaterialSwitch before. I talked with @tobiasKaminsky he found the switch is too big that's why I used this approach and scaleX and scaleY. 

I would also go with MaterialSwitch. @tobiasKaminsky What do you think?

File: app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java [hunk 6]
```diff
@@ -789,4 +793,24 @@ public void setLastSelectedMediaFolder(@NonNull String path) {
...
     }
+
+    @Override
+    public void setTwoWaySyncStatus(boolean value) {
+        preferences.edit().putBoolean(PREF__TWO_WAY_STATUS, value).apply();
+    }
+
+    @Override
+    public boolean isTwoWaySyncEnabled() {
+        return preferences.getBoolean(PREF__TWO_WAY_STATUS, true);
```
[the remark this answers is not in the document]
Comment (alperozturk96) [reply app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java:804]: By default, this feature was disabled, but I changed it because users can still enable sync for a specific folder from the folder details without realizing they also need to activate this feature in settings. Therefore, I set the default to true.

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -115,6 +115,19 @@
...
+    <string name="two_way_sync_interval_15_min">15 minutes</string>
+    <string name="two_way_sync_interval_30_min">30 minutes</string>
+    <string name="two_way_sync_interval_45_min">45 minutes</string>
+    <string name="two_way_sync_interval_1_hour">1 hour</string>
+    <string name="two_way_sync_interval_2_hours">2 hours</string>
+    <string name="two_way_sync_interval_4_hours">4 hours</string>
+    <string name="two_way_sync_interval_6_hours">6 hours</string>
+    <string name="two_way_sync_interval_8_hours">8 hours</string>
+    <string name="two_way_sync_interval_12_hours">12 hours</string>
+    <string name="two_way_sync_interval_24_hours">24 hours</string>
```
[the remark this answers is not in the document]
Comment (alperozturk96) [reply app/src/main/res/values/strings.xml]: https://github.com/nextcloud/android/pull/13884/commits/c11d7e2e74ff459bb0bc169ad7a949a941cfe203

File: app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java [hunk 8]
```diff
@@ -102,6 +102,10 @@ public final class AppPreferencesImpl implements AppPreferences {
     private static final String PREF__STORAGE_PERMISSION_REQUESTED = "storage_permission_requested";
     private static final String PREF__IN_APP_REVIEW_DATA = "in_app_review_data";
 
+    private static final String PREF__TWO_WAY_STATUS = "two_way_sync_status";
+    private static final String PREF__RESET_TWO_WAY_SYNC_TIMESTAMP = "reset_two_way_sync_timestamp";
```
[the remark this answers is not in the document]
Comment (alperozturk96) [reply app/src/main/java/com/nextcloud/client/preferences/AppPreferencesImpl.java]: https://github.com/nextcloud/android/pull/13884/commits/781449e45d7dd7734e0629586d2e9cc01921c533

File: app/src/main/res/layout/internal_two_way_sync_layout.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (alperozturk96) [reply app/src/main/res/layout/internal_two_way_sync_layout.xml]: https://github.com/nextcloud/android/pull/13884/commits/6d5e15fc0c46ced298d489cfc7d8ebe4c7244e84

File: app/src/main/res/layout/internal_two_way_sync_layout.xml [hunk 9]
```diff
@@ -5,14 +5,60 @@
...
+            android:text="@string/prefs_two_way_sync_switch_title"
+            android:textSize="@dimen/txt_size_16sp"
+            android:layout_weight="9"
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"/>
+
+        <com.google.android.material.materialswitch.MaterialSwitch
+            android:id="@+id/twoWaySyncToggle"
+            android:scaleX="0.7"
+            android:scaleY="0.7"
```
[the remark this answers is not in the document]
Comment (alperozturk96) [reply app/src/main/res/layout/internal_two_way_sync_layout.xml]: https://github.com/nextcloud/android/pull/13884/commits/6d5e15fc0c46ced298d489cfc7d8ebe4c7244e84

File: app/src/main/java/com/nextcloud/model/DurationOption.kt (hunk 4 above)
Comment (solonovamax) [reply app/src/main/java/com/nextcloud/model/DurationOption.kt]: you *can* do `.toString()` on it, which returns it in the format `1h 0m 30.340s`, however it does not support i18n. if you want to do some i18n stuff, you could use `.toComponents { days, hours, minutes, seconds, nanoseconds -> }`

File: app/src/main/java/com/nextcloud/client/jobs/BackgroundJobManagerImpl.kt [hunk 10]
```diff
@@ -698,12 +698,13 @@ internal class BackgroundJobManagerImpl(
         )
     }
 
-    override fun scheduleInternal2WaySync() {
+    override fun scheduleInternal2WaySync(intervalMinutes: Long) {
```
Comment (tobiasKaminsky) [inline app/src/main/java/com/nextcloud/client/jobs/BackgroundJobManagerImpl.kt:701]: If appPreferences.isTwoWaySyncEnabled == false then we should not schedule it at all?

File: app/src/main/java/com/owncloud/android/ui/activity/InternalTwoWaySyncActivity.kt [hunk 11]
```diff
@@ -8,23 +8,91 @@
...
+            15.minutes to getString(R.string.two_way_sync_interval_minutes, "15"),
+            30.minutes to getString(R.string.two_way_sync_interval_minutes, "30"),
+            45.minutes to getString(R.string.two_way_sync_interval_minutes, "45"),
+            1.hours to getString(R.string.two_way_sync_interval_hour, "1"),
+            2.hours to getString(R.string.two_way_sync_interval_hours, "2"),
+            4.hours to getString(R.string.two_way_sync_interval_hours, "4"),
+            6.hours to getString(R.string.two_way_sync_interval_hours, "6"),
+            8.hours to getString(R.string.two_way_sync_interval_hours, "8"),
+            12.hours to getString(R.string.two_way_sync_interval_hours, "12"),
+            24.hours to getString(R.string.two_way_sync_interval_hours, "24")
```
Comment (tobiasKaminsky) [inline app/src/main/java/com/owncloud/android/ui/activity/InternalTwoWaySyncActivity.kt:52]: I fear this approach will not work for all languages.
As far as I know there are languages that e.g. have an other "wording" for 1, 2 or 3+.

But I am unsure how to do it better.
We can probably later enhance it.

File: app/src/main/res/xml/preferences.xml [hunk 12]
```diff
@@ -9,133 +9,145 @@
   ~ SPDX-License-Identifier: GPL-2.0-only AND (AGPL-3.0-or-later OR GPL-2.0-only)
 -->
 <PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android"
-				  android:key="preference_screen">
+    android:key="preference_screen">
```
Comment (tobiasKaminsky) [inline app/src/main/res/xml/preferences.xml:12]: Please avoid such refactoring in a bugfix PR.
Instead create a new one. 
This makes it very hard to review.

**Review by tobiasKaminsky — APPROVED** [review]

File: app/src/main/res/xml/preferences.xml (hunk 12 above)
Comment (alperozturk96) [reply app/src/main/res/xml/preferences.xml:12]: I tried to add a switch there first and that's why this happened. I can undo the changes

File: app/src/main/java/com/owncloud/android/ui/activity/InternalTwoWaySyncActivity.kt (hunk 11 above)
Comment (ZetaTom) [reply app/src/main/java/com/owncloud/android/ui/activity/InternalTwoWaySyncActivity.kt:52]: We could use [quantity strings](https://developer.android.com/guide/topics/resources/string-resource#Plurals). I'm not sure if we'll be able to handle all edge cases, but this seems to be the "official" approach.

File: app/src/main/res/xml/preferences.xml (hunk 12 above)
Comment (alperozturk96) [reply app/src/main/res/xml/preferences.xml]: https://github.com/nextcloud/android/pull/13884/commits/5fb1a0ebc389e43feaa1572671e36a25516cf8c7

File: app/src/main/java/com/nextcloud/client/jobs/BackgroundJobManagerImpl.kt (hunk 10 above)
Comment (alperozturk96) [reply app/src/main/java/com/nextcloud/client/jobs/BackgroundJobManagerImpl.kt:701]: https://github.com/nextcloud/android/pull/13884/commits/5aca575972d9173657a45d44cd340edbf66e8219

**Review by ZetaTom — COMMENTED** [review]
Everything works as expected on my device.

Please see my previous [comment](https://github.com/nextcloud/android/pull/13884#discussion_r1822205683). If we agree, you will have my approval.

File: app/src/main/java/com/owncloud/android/ui/activity/InternalTwoWaySyncActivity.kt (hunk 11 above)
Comment (alperozturk96) [reply app/src/main/java/com/owncloud/android/ui/activity/InternalTwoWaySyncActivity.kt]: https://github.com/nextcloud/android/pull/13884/commits/7ec78948a7ecae5d207dd4b9118fab91286da068

**Comment by alperozturk96** [comment]
@tobiasKaminsky

Test regarding three-dot-menu visibility

https://github.com/user-attachments/assets/1431eb74-199f-42d2-b24d-90547d36d62d

**Comment by Hannes-Schniz** [comment]
Looking forward to this PR currently sync is not usable for me due to that!
